### ISSUE TITLE
Dart: optimize Locals.get; it is showing in profiles a lot

### ISSUE THREAD

**Issue by yjbanov** [angular/angular#5970]
`Locals` defeats `dart2js` type inference in that it expects arbitrary map type to be passed as constructor argument. In reality we only ever use maps with string keys, which could be reduced to a plain JS object.

Here's what e.g. `Locals.get` compiles to:

```js
      $get$1: function(_, $name) {
        var t1;
        if (J.containsKey$1$x(this.current, $name))
          return J.$index$asx(this.current, $name);
        t1 = this.parent;
        if (t1 != null)
          return t1.$get$1(0, $name);
        throw H.wrapException(new L.BaseException("Cannot find '" + H.S($name) + "'"));
      }
```

NOTE: you may see different JS code in smaller apps because smaller apps tend to use fewer kinds of maps that could lead to better inference. The problem happens in larger apps.

Here's the Dart code for `Local` that we `ts2dart` from our TypeScript code:

```dart
library angular2.src.core.change_detection.parser.locals;

import "package:angular2/src/facade/lang.dart" show isPresent;
import "package:angular2/src/facade/exceptions.dart" show BaseException;
import "package:angular2/src/facade/collection.dart"
    show ListWrapper, MapWrapper;

class Locals {
  Locals parent;
  Map<dynamic, dynamic> current;
  Locals(this.parent, this.current) {}
  bool contains(String name) {
    if (this.current.containsKey(name)) {
      return true;
    }
    if (isPresent(this.parent)) {
      return this.parent.contains(name);
    }
    return false;
  }

  dynamic get(String name) {
    if (this.current.containsKey(name)) {
      return this.current[name];
    }
    if (isPresent(this.parent)) {
      return this.parent.get(name);
    }
    throw new BaseException('''Cannot find \'${ name}\'''');
  }

  void set(String name, dynamic value) {
    // TODO(rado): consider removing this check if we can guarantee this is not

    // exposed to the public API.

    // TODO: vsavkin maybe it should check only the local map
    if (this.current.containsKey(name)) {
      this.current[name] = value;
    } else {
      throw new BaseException(
          '''Setting of new keys post-construction is not supported. Key: ${ name}.''');
    }
  }

  void clearValues() {
    MapWrapper.clearValues(this.current);
  }
}
```

/cc @sigmundch @hterkelsen @rakudrama

**Comment by rakudrama** [comment]
1. Make `Locals.current` final, private, or both.

Public non-final fields can appear to be overwritten by random other code that also has fields with the same name. This prevents type inference from figuring out the type. The probability of a such a type inference accident increases with code size.

The fact that `this.current` occurs twice in succession in the generated code which indicates that dart2js thinks it might have changed between uses.

If you can't do that, do `var current = this.current` and use the temporary.

2. If you are using --trust-type-annotations, is there a more specific type you could use in the declaration of `current`?

3. Call `contains` only when the result is `null`.
(The Map API really does need a __lookup__ method that takes a default so you can pass in a token to detect missing elements without two calls.)

```Dart
    var current = this.current;
    var value = current[name];
    if (value != null || current.contains(name)) return value;
   ...
```

**Comment by yjbanov** [comment]
I was actually thinking of removing the `current` constructor parameter completely and making `Locals.current` private, final and instantiate it to something as close to a plain JS object as possible (perhaps even use JS interop). Given that we have `get`/`set` in the class already, there's no need to leave `current` so wide open to intrusion.

**Comment by tbosch** [comment]
The locals map will soon go away with the new codegen in #6270.

**Comment by tbosch** [comment]
Closing this as this will be automatically solved via #6270
